### PR TITLE
✨Configuration assembly app layer

### PR DESCRIPTION
**Pull request by muchzill4**
### Description

Adding mbed_app.json layer on top of the existing config build process.

Works against blinky, but explodes with cloud client example. This is due to bootloader overrides which are not defined in targets.json base "config", yet seem to exist in "target_overrides" of mbed_app.json. Fixing that is subject to another PR.

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_build/_internal/config/assemble_build_config.py [hunk 1]
```diff
@@ -21,16 +22,22 @@ def assemble_config(mbed_target: str, mbed_program_directory: Path) -> Config:
     """
     target_source = Source.from_target(mbed_target, mbed_program_directory)
     mbed_lib_files = find_files("mbed_lib.json", mbed_program_directory)
-    return _assemble_config(target_source, mbed_lib_files)
+    mbed_program = MbedProgram.from_existing_local_program_directory(mbed_program_directory)
```
Comment (rosiemiller2) [inline mbed_build/_internal/config/assemble_build_config.py:25]: I think these method names have changed in the latest release of mbed-project

File: mbed_build/_internal/config/assemble_build_config.py [hunk 2]
```diff
@@ -21,16 +22,22 @@ def assemble_config(mbed_target: str, mbed_program_directory: Path) -> Config:
...
+    mbed_app_file = mbed_program.metadata.config_file
```
Comment (rosiemiller2) [inline mbed_build/_internal/config/assemble_build_config.py:26]: I think the name 'metadata' might have changed too...

Comment (muchzill4) [reply mbed_build/_internal/config/assemble_build_config.py]: https://github.com/ARMmbed/mbed-build/pull/35/commits/98981df2f289be03578d919394e81a9e86abf4cf

File: mbed_build/_internal/config/assemble_build_config.py (hunk 1 above)
Comment (muchzill4) [reply mbed_build/_internal/config/assemble_build_config.py]: https://github.com/ARMmbed/mbed-build/pull/35/commits/98981df2f289be03578d919394e81a9e86abf4cf

File: mbed_build/_internal/config/source.py [hunk 3]
```diff
@@ -38,21 +38,43 @@ def from_mbed_lib(cls, file: Path, target_labels: Iterable[str]) -> "Source":
...
+    @classmethod
+    def from_mbed_app(cls, file: Path, target_labels: Iterable[str]) -> "Source":
+        """Build Source from mbed_app.json file.
+
+        Args:
+            file: Path to mbed_app.json file
+            target_labels: Labels for which "target_overrides" should apply
+        """
+        file_contents = json.loads(file.read_text())
+        return cls.from_file_contents(
```
Comment (rosiemiller2) [inline mbed_build/_internal/config/source.py:57]:
```suggestion
        return cls.from_json(
```  ?

Comment (rosiemiller2) [reply mbed_build/_internal/config/source.py:57]: Actually, i think your original way was better. I retract my suggestion.

**Pull request merged**